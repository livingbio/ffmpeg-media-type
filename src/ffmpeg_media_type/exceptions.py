class FfmpegMediaTypeError(Exception):
    """
    Base class for exceptions in this module.
    """
