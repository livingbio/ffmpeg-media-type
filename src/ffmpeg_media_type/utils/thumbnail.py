from pathlib import Path

from .hotfix_webp import hotfix_animate_webp
from .shell import call, create_temp_filename


def generate_thumbnail(
    video_path: str | Path, suffix: str = ".png", *, width: int = 320, height: int = -1, time_offset: float = 0
) -> str:
    """
    Generate a thumbnail from a video file at a specified time offset.

    Args:
        video_path: the path to the video file
        suffix: the suffix of the generated thumbnail
        width: the width of the generated thumbnail
        height: the height of the generated thumbnail
        time_offset: the time offset in seconds to generate the thumbnail

    Raises:
        FfmpegMediaTypeError: If the ffmpeg command fails.

    Returns:
        the path to the generated thumbnail
    """

    video_path = hotfix_animate_webp(video_path)

    thumbnail_path = create_temp_filename(suffix)
    ffmpeg_cmd = ["ffmpeg"] + [
        "-y",  # Overwrite output file if it exists
        "-i",
        str(video_path),  # Input video path
        "-ss",
        str(time_offset),  # Time offset (seek to the specified position)
        "-vframes",
        "1",  # Number of frames to output
        "-vf",
        f"scale={width}:{height}",  # Thumbnail size (width: 320, height: proportional)
        "-q:v",
        "2",  # Quality (2 - high, 5 - low)
        thumbnail_path,  # Output thumbnail path
    ]

    call(ffmpeg_cmd)

    return thumbnail_path
