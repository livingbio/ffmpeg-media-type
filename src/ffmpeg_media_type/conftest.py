from pathlib import Path

import pytest
from syrupy.assertion import SnapshotAssertion
from syrupy.extensions.amber import AmberSnapshotExtension
from syrupy.location import PyTestLocation

from .utils.ffmpeg import get_ffmpeg_version


@pytest.fixture(scope="session")
def ffmpeg_version() -> str:
    return get_ffmpeg_version("minor")


class DifferentDirectoryExtension(AmberSnapshotExtension):
    @classmethod
    def dirname(cls, *, test_location: "PyTestLocation") -> str:
        version = get_ffmpeg_version("minor")
        return str(Path(test_location.filepath).parent / "__snapshots__" / version)


@pytest.fixture
def snapshot_ffmpeg(snapshot: SnapshotAssertion) -> SnapshotAssertion:
    return snapshot.use_extension(DifferentDirectoryExtension)
