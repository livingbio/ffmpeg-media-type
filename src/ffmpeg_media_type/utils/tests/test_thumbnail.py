from pathlib import Path

import pytest
from syrupy.assertion import SnapshotAssertion

from ...conftest import sample_test_media_files
from ...exceptions import FfmpegMediaTypeError
from ..thumbnail import generate_thumbnail


@pytest.mark.parametrize("case", sample_test_media_files())
def test_thumbnail(case: Path, snapshot: SnapshotAssertion) -> None:
    try:
        generate_thumbnail(case)
        assert snapshot == True
    except FfmpegMediaTypeError:
        assert snapshot == False
