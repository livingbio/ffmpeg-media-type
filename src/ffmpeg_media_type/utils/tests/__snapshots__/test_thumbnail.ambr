# serializer version: 1
# name: test_thumbnail[10257640.mp4]
  True
# ---
# name: test_thumbnail[188794898-eeb81e4b-b33a-4bdf-85a8-3b4b8460b88a.png]
  True
# ---
# name: test_thumbnail[201432913186_fish.gif]
  True
# ---
# name: test_thumbnail[4ox1IBJxK3cVd0uJA6gOtceSP3g.jpg]
  True
# ---
# name: test_thumbnail[9b5174019e8c9bec190903528ff6d0392a2d6a9b.mp4]
  True
# ---
# name: test_thumbnail[9c7ee468-24a3-4930-9832-73f0f6001ab1.webp]
  True
# ---
# name: test_thumbnail[Cle9pQ5.jpg]
  True
# ---
# name: test_thumbnail[SampleAudio_0.4mb.mp3]
  False
# ---
# name: test_thumbnail[SampleVideo_1280x720_1mb.mp4]
  True
# ---
# name: test_thumbnail[animated-webp.webp]
  False
# ---
# name: test_thumbnail[cd89b1bf161582d55161f5ba94d03cc7728825c2.jpe]
  True
# ---
# name: test_thumbnail[lionic_vp8x_1109.webp]
  True
# ---
# name: test_thumbnail[maxresdefault.jpg-sdafdeagwdsadf]
  True
# ---
# name: test_thumbnail[mp3-with-cover.mp3]
  True
# ---
# name: test_thumbnail[overlay.png]
  True
# ---
# name: test_thumbnail[silent_video.mp40]
  True
# ---
# name: test_thumbnail[silent_video.mp41]
  True
# ---
